ignore patterns defined in dockerignore

Luukvdm opened the pull request:
<!-- Type of change
Please label this PR with one of the existing labels, depending on the scope of your change
-->

## What does this PR do?

Respect `.dockerignore` files similar to how the Docker CLI [handles them](https://github.com/docker/cli/blob/master/cli/command/image/build/dockerignore.go#L14).

## Why is it important?

Prevents unnecessary large docker images and mirrors behavior of the Docker CLI.

## Related issues

- Closes #1486

mdelapenya commented on container.go:
```diff
@@ -188,7 +190,17 @@ func (c *ContainerRequest) GetContext() (io.Reader, error) {
+	var excluded []string
+	if f, err := os.Open(filepath.Join(abs, ".dockerignore")); err != nil {
+		var err error
+		excluded, err = ignorefile.ReadAll(f)
```
To avoid a new dependency, I'd also copy the function in here: https://github.com/moby/patternmatcher/blob/main/ignorefile/ignorefile.go

Probably into the `internal/testcontainersdocker` package.

What are your thoughts on this?

mdelapenya replied:
Exactly, that's what I think too: it's basically parsing the dockerignore file. Only if you agree, let's copy that, but as you said, it's part of the indirect deps already. Whatever is your preference I'm fine with it.

Luukvdm replied:
I think I prefer leaving it as is. Because the dependency is small and its easier to track upstream changes this way.

mdelapenya commented on container.go:
```diff
@@ -188,7 +190,17 @@ func (c *ContainerRequest) GetContext() (io.Reader, error) {
+		var err error
+		excluded, err = ignorefile.ReadAll(f)
+		if err != nil {
+			return nil, fmt.Errorf("error reading .dockerignore: %w", err)
```
Do we want to raise an error here and fail our logic, or simply log the error and continue?

Luukvdm replied:
At this point in the code the dockerignore file is read, and any errors are coming from the parsing of the file. I feel like if this ever fails its better to return the error and let the user fix the problem.

mdelapenya commented:
@Luukvdm there is a few tests failing. Please see https://github.com/testcontainers/testcontainers-go/actions/runs/6448106348/job/17580097884?pr=1725

They are all related to the `.dockerignore` new feature we added in this PR, very likely because the file does not exist for the tests. I think we are missing a check for the file existence 🤔 

Thankfully we have tests! 💪

Luukvdm commented:
> @Luukvdm there is a few tests failing. Please see https://github.com/testcontainers/testcontainers-go/actions/runs/6448106348/job/17580097884?pr=1725
> 
> They are all related to the `.dockerignore` new feature we added in this PR, very likely because the file does not exist for the tests. I think we are missing a check for the file existence 🤔
> 
> Thankfully we have tests! 💪

Thank god for automated testing 😄
I made a silly mistake in the error handling. I fixed it and added a regression test just in case. Do you agree with the changes? 
Also wasn't sure about code styling/ order and added the function under the `GetContext` function.

mdelapenya reviewed (approved):
LGTM, thanks for your contribution!
